### ISSUE TITLE
Text preprocessing

### ISSUE THREAD

**Issue by hp0404** [MaartenGr/BERTopic#40]
Hi! Thanks for developing this awesome library! 

I have a question regarding text preprocessing. 

From what I understand, the model takes `List[str]` as an input - basically a list of fulltext documents. 

*But do we need to preprocess texts somehow before passing it into the model?*

With LDA, I usually preprocess texts (tokenize, lemmatize, remove stopwords, create n-grams, etc.) before running models. But since we're dealing with word embeddings, keeping all words in their original form is important for the context, right? 

So I'm not sure how to proceed, should I use list of preprocessed words as an input, or leave texts untouched, or something in between (keeping text as a string but without stopwords, etc.)?

**Comment by MaartenGr** [comment]
In general, no, you do not need to preprocess your data. Like you said, keeping the original structure of the text is especially important for transformer-based models to understand the context. 

However, there are exceptions to this. For example, if you were to have scraped documents with a lot of html tags, then it might be beneficial to remove those as they do not provide any interesting context. 

If you have paragraphs in a document, then it might be worthwhile to split up the paragraphs in order to more precisely extract the correct topics.

**Comment by hp0404** [comment]
Great, thanks!

**Comment by KuangHao95** [comment]
Hi, I think you still need to remove stop words, since there is no longer a `stop_words` parameter in BERTopic().

**Comment by hp0404** [comment]
What about 'keeping the original structure of the text'? @KuangHao95

**Comment by MaartenGr** [comment]
Although the `stop_words` parameter was removed in newer versions, you are still able to remove stopwords by using the CountVectorizer!

```python
from bertopic import BERTopic
from sklearn.feature_extraction.text import CountVectorizer

vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words="english")
topic_model = BERTopic(vectorizer_model=vectorizer_model)
```

The `stop_words` parameter simply exposed the CountVectorizer parameter. It was removed because at some point I could expose all parameters of HDBSCAN, UMAP, and CountVectorizer into BERTopic which would make the API ambiguous.

**Comment by Ayush1702** [comment]
> In general, no, you do not need to preprocess your data. Like you said, keeping the original structure of the text is especially important for transformer-based models to understand the context.
> 
> However, there are exceptions to this. For example, if you were to have scraped documents with a lot of html tags, then it might be beneficial to remove those as they do not provide any interesting context.
> 
> If you have paragraphs in a document, then it might be worthwhile to split up the paragraphs in order to more precisely extract the correct topics.

Well in my case, working on a custom dataset, I saw stopwords landing in my topic results as Salient terms, so should I specify stopwords in CountVectorizer?

**Comment by MaartenGr** [comment]
Stopwords typically appear if you have very few documents or if the documents are quite short. So either adding more documents or using longer documents might solve this issue. Obviously, this is often not possible. In that case, you can indeed specify stopwords in the CountVectorizer. You can find a bit more about that [here](https://maartengr.github.io/BERTopic/tutorial/models/models.html).

**Comment by Ayush1702** [comment]
> Stopwords typically appear if you have very few documents or if the documents are quite short. So either adding more documents or using longer documents might solve this issue. Obviously, this is often not possible. In that case, you can indeed specify stopwords in the CountVectorizer. You can find a bit more about that [here](https://maartengr.github.io/BERTopic/tutorial/models/models.html).

I do have 23k documents but the length varies, some are long while some are short, so it's not exactly consistent like tweets or the newsgroup dataset. So maybe I'm better off specifying only fewer stopwords

**Comment by fishfree** [comment]
@MaartenGr BTW: how to add my custom stopwords, please?

**Comment by MaartenGr** [comment]
@fishfree You can use [this](https://maartengr.github.io/BERTopic/tutorial/models/models.html) part of the documentation to add custom stopwords. Simply replace `"english"` with a list of stop words and you are good to go!

**Comment by jasonmorkel** [comment]
@MaartenGr is seems this link is no longer valid - do you have a new link to documentation on adding custom stopwords?

**Comment by MaartenGr** [comment]
@jasonmorkel Yes! You can use [this](https://maartengr.github.io/BERTopic/getting_started/models/models.html) link with a description of using your own stop words. 

The documentation structure is likely changing with the upcoming update, so I will also post the method of doing so below:

```python
from bertopic import BERTopic
from sklearn.feature_extraction.text import CountVectorizer

vectorizer_model = CountVectorizer(stop_words=my_list_of_stopwords)
topic_model = BERTopic(vectorizer_model=vectorizer_model)
```

Note that the `CountVectorizer` processes the documents **after** they are clustered which means that you can use it to clean the documents and optimize your topic representations.

**Comment by jasonmorkel** [comment]
@MaartenGr Thank you! 

For my specific dataset, the topics became a lot more interpretable by using stopwords. 

Amazing results!!

**Comment by jacobceles** [comment]
Hi @MaartenGr 

Please correct me if I am wrong but if I understand right, the [all-mpnet-base-v2](https://www.sbert.net/docs/pretrained_models.html) model has a maximum length limit of tokens as 384. I am dealing with long text, so I believe one approach would be to cut the longer texts into multiple with 384 tokens each. Is this handled within the library or do I have to write separate code for this?

**Comment by MaartenGr** [comment]
@jacobceles This is currently not handled within BERTopic as there are many variations of embedding models to be used and it is not necessary for all models (e.g., TF-IDF representations). So you would indeed have to separate the text yourself. Typically, you can get quite far already with the `sent_tokenize` function in [NLTK](https://www.nltk.org/api/nltk.tokenize.html).

**Comment by eyast** [comment]
latest link posted by @MaartenGr  leads to a 404?

**Comment by MaartenGr** [comment]
@eyast I just tried out the link and it does not lead to a 404 to me. The `sent_tokenize` function in NLTK works as follows:

```python
>>> from nltk.tokenize import sent_tokenize
>>> sent_tokenize(s)
```

**Comment by eyast** [comment]
> @fishfree You can use [this](https://maartengr.github.io/BERTopic/tutorial/models/models.html) part of the documentation to add custom stopwords. Simply replace `"english"` with a list of stop words and you are good to go!

@MaartenGr - I was referring to this one, apologies

**Comment by MaartenGr** [comment]
@eyast No problem, I updated the link!

**Comment by Sinandiao** [comment]
å

> [@jacobceles](https://github.com/jacobceles) This is currently not handled within BERTopic as there are many variations of embedding models to be used and it is not necessary for all models (e.g., TF-IDF representations). So you would indeed have to separate the text yourself. Typically, you can get quite far already with the `sent_tokenize` function in [NLTK](https://www.nltk.org/api/nltk.tokenize.html).

Hi! I searched the entire web and couldn't find a better way to handle long texts. Are there any solutions for the long text which is more than 512 tokens?Is chunk the only solution? Thank you.

**Comment by Sinandiao** [comment]
> [@jacobceles](https://github.com/jacobceles) This is currently not handled within BERTopic as there are many variations of embedding models to be used and it is not necessary for all models (e.g., TF-IDF representations). So you would indeed have to separate the text yourself. Typically, you can get quite far already with the `sent_tokenize` function in [NLTK](https://www.nltk.org/api/nltk.tokenize.html).

Hi! I searched the entire web and couldn't find a better way to handle long texts. Are there any solutions for the long text which is more than 512 tokens?Is chunk the only solution? Thank you.

**Comment by MaartenGr** [comment]
@Sinandiao Many embedding models these day can handle texts longer than 512 tokens. That's only a limit for some of these models. For instance, [gte-large-en-v1.5](https://huggingface.co/Alibaba-NLP/gte-large-en-v1.5) can handle contexts of 8k tokens. 

That said, for topic modeling (and especially cluster-based models) it makes sense to want to split your documents up into smaller chunks. I'm doing that myself often and it works quite well.

**Comment by MaartenGr** [comment]
@Sinandiao Personally, I tend to track these chunks by their membership to a given document.

**Comment by Sinandiao** [comment]
> [@Sinandiao](https://github.com/Sinandiao) Personally, I tend to track these chunks by their membership to a given document.

Thanks a lot for your reply! I really appreciate it.
